server/plugin/oauth.go

```diff
@@ -0,0 +1,90 @@
+package plugin
+
+import (
+	"sync"
+)
+
+type OAuthCompleteEvent struct {
+	UserID string
+	Err    error
+}
+
+type OAuthBroker struct {
+	sendOAuthCompleteEvent func(event OAuthCompleteEvent)
+
+	lock              sync.RWMutex // Protects closed and pingSubs
+	closed            bool
+	oauthCompleteSubs map[string][]chan error
+	mapCreate         sync.Once
+}
+
+func NewOAuthBroker(sendOAuthCompleteEvent func(event OAuthCompleteEvent)) *OAuthBroker {
+	return &OAuthBroker{
+		sendOAuthCompleteEvent: sendOAuthCompleteEvent,
+	}
+}
+
+func (ob *OAuthBroker) SubscribeOAuthComplete(userID string) <-chan error {
+	ob.lock.Lock()
+	defer ob.lock.Unlock()
+
+	ob.mapCreate.Do(func() {
+		ob.oauthCompleteSubs = make(map[string][]chan error)
+	})
+
+	ch := make(chan error, 1)
+	ob.oauthCompleteSubs[userID] = append(ob.oauthCompleteSubs[userID], ch)
+
+	return ch
+}
+
+func (ob *OAuthBroker) UnsubscribeOAuthComplete(userID string, ch <-chan error) {
+	ob.lock.Lock()
+	defer ob.lock.Unlock()
+
+	for i, sub := range ob.oauthCompleteSubs[userID] {
+		if sub == ch {
+			close(sub)
+			ob.oauthCompleteSubs[userID] = append(ob.oauthCompleteSubs[userID][:i], ob.oauthCompleteSubs[userID][i+1:]...)
+			break
+		}
+	}
+}
+
+func (ob *OAuthBroker) publishOAuthComplete(userID string, err error, fromCluster bool) {
+	ob.lock.Lock()
+	defer ob.lock.Unlock()
+
+	if ob.closed {
+		return
+	}
+
+	for _, userSubs := range ob.oauthCompleteSubs {
+		for _, sub := range userSubs {
+			// non-blocking send
+			select {
+			case sub <- err:
+			default:
+			}
+		}
+	}
+
+	if !fromCluster {
+		ob.sendOAuthCompleteEvent(OAuthCompleteEvent{UserID: userID, Err: err})
+	}
+}
```

Comment:
I'm not sure I understand what's going on here. Why do the subscribed channels not need to know what the user id is for this event?

Reply:
Good catch :+1: Fixed in [`21ad0be` (#541)](https://github.com/mattermost/mattermost-plugin-github/pull/541/commits/21ad0be7d78cb6708e809e0690f39b213dcb0142), which is in https://github.com/mattermost/mattermost-plugin-github/pull/541.